### PR TITLE
Add `XLATrainer` subclass to implement UNet3D training in PyTorch/XLA

### LINKED ISSUE #8: Refactor `training` module for readability and extensibility with `torch_xla`
We hope to refactor the `training` module into a more readable and extensible `trainer` package.

The `trainer` package uses polymorphism (the base class is `UNet3DTrainer` & the concrete sub-classes are `CUDATrainer` and `XLATrainer`) to toggle between training with CUDA and training with PT-XLA during runtime.

For more details, see b/224290413

### PR DESCRIPTION
**Pull request by thisisalbertliang**
Resolve #8 (partially)

Related to:
- `UNet3DTrainer` PR: https://github.com/thisisalbertliang/training/pull/9

We hope to refactor the `training` module into a more readable and extensible `trainer` package.

The `trainer` package uses polymorphism (the base class is `UNet3DTrainer` & the concrete sub-classes are `CUDATrainer` and `XLATrainer`) to toggle between training with CUDA and training with PT-XLA during runtime.

For more details, see b/224290413

### CODE REVIEW

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        val_loader,
+        loss_fn,
+        score_fn,
+        device,
+        callbacks,
+    ) -> None:
+        super().__init__(
+            flags, model, train_loader, val_loader, loss_fn, score_fn, device, callbacks
+        )
+        assert isinstance(self.train_loader, pl.MpDeviceLoader), (
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/xla_trainer.py:27]: I expect this object is created by `get_trainer()` in this [PR](https://github.com/thisisalbertliang/training/pull/12).

What's the behavior for a single-core job?

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        # Setup train sampler
+        self.train_sampler = self.train_loader._loader.sampler
+
+        # Get hardware type
+        self.hw_type = xm.xla_device_hw(self.device)
+
+    def forward_pass(self, image: torch.Tensor, label: torch.Tensor) -> torch.Tensor:
+        """Overrides UNet3DTrainer.forward_pass"""
+        with torch_xla.amp.autocast(enabled=self.flags.amp):
+            output = self.model(image)
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/xla_trainer.py]: To be clear, can image be a list of images? (i.e. batch size > 1)

File: image_segmentation/pytorch/runtime/trainer/test/test_xla_trainer.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        flags = Namespace(
+            amp=False,
+            ga_steps=1,
+            layout="NCDHW",
+            include_background=False,
+            batch_size=self.batch_size,
+            input_shape=self.image_size,
+            val_input_shape=self.image_size,
+            benchmark=False,
+            num_workers=8,
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/trainer/test/test_xla_trainer.py]: I'd test 8 and 1

**Review by miladm — COMMENTED** [review]
Thanks @thisisalbertliang. I added a few comments below.

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py (hunk 1 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/xla_trainer.py:27]: Yes. `XLATrainer` should only be created from `get_trainer()`

For single core training, the train/val loaders should also be `pl.MpDeviceLoader` objects, hence why I made these python assertions. I expect `pl.MpDeviceLoader` to also bring performance improvement on a single-core jobs due to its batch prefetching optimizations.

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py (hunk 2 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/xla_trainer.py]: No. `image` is always a `torch.Tensor` object, as indicated in the type annotation.

To over-communicate a bit, the shape of `image` is `(batch_size, channel_size, x_size, y_size, z_size)`. When batch size > 1, `image` simply gets thicker in the `batch_size` dimension. There is no python list here at all.

File: image_segmentation/pytorch/runtime/trainer/test/test_xla_trainer.py (hunk 3 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/test/test_xla_trainer.py]: Thanks for the feedback. I just added additional tests to test for both `num_workers=1` and `num_workers=8`.

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py (hunk 2 above)
Comment (will-cromar) [reply image_segmentation/pytorch/runtime/trainer/xla_trainer.py]: nit: if the input is a batch of images, can you change the name to `images`? I would have assumed a shape of `(channel_size, x_size, y_size, z_size)` given the name `image`.

**Review by will-cromar — CHANGES_REQUESTED** [review]

File: image_segmentation/pytorch/runtime/trainer/xla_trainer.py (hunk 2 above)
Comment (thisisalbertliang) [reply image_segmentation/pytorch/runtime/trainer/xla_trainer.py]: Sure. I have changed all ocurrences of `image` and `label` (in `xla_trainer.py`, `test_xla_trainer.py`, `unet3d_trainer.py`, `cuda_trainer.py`) to `images` and `labels`.

**Review by will-cromar — APPROVED** [review]
